WordPress/wordpress-playground
# Avoid partial munmap memory leak

## What is this PR doing?

This PR works around a PHP memory leak that can quickly lead to an OOM condition.

## What problem is it solving?

This PR fixes #1128 -- an OOM condition due to memory leak.

## How is the problem addressed?

This PR adds custom PHP memory storage handlers to stop PHP from attempting to partial unmap memory via the `munmap()` function. Emscripten allocators do not support using `munmap()` to partially unmap memory. They fail for partial unmaps, and when attempts fail, memory is leaked because PHP thinks the memory is free while Emscripten libs consider it allocated.

The custom memory storage handlers take over responsibility for allocating and freeing aligned memory chunks. The handlers use `posix_memalign()` and `free()` instead of `mmap()` and `munmap()`, and this appears to resolve the problem.

The handlers are added as part of a PHP extension, which is not perfect because allocation can happen before the extension is initialized. But the amount of allocation before the workaround is applied should be quite small.

## Testing Instructions

Before applying this patch:
Run [this test script](https://gist.github.com/brandonpayton/934d96d76b96557e94ad6983aaffa09f) and verify that it OOMs.

After applying this patch:
Rerun the test script and verify that it completes after 1000 interations.

## Remaining work

- [x] Compare performance with php-wasm builds prior to this patch and reconsider if 20% slower.
- [x] Find out whether it is sufficient to add the workaround during the PHP module init phase rather than during each request init
- [x] Address minor TODO comments in wasm_memory_storage.c
- [x] See if we can run the test script as part of tests under `packages/php-wasm/node/src/test`
